src/RevisionPopupWidget.js

```diff
@@ -48,29 +51,52 @@ RevisionPopupWidget.prototype.show = function ( data, $target ) {
 		contribsUrl = mw.util.getUrl( `Special:Contributions/${username}` ),
 		// We typically link to Special:Contribs for IPs.
 		userPageUrl = isIP ? contribsUrl : mw.util.getUrl( `User:${data.username}` ),
-		userLinks = `
-			<a href="${userPageUrl}">${username}</a>
-			${mw.msg( 'parentheses-start' )}<a href="${mw.util.getUrl( `User talk:${username}` )}">${mw.msg( 'talkpagelinktext' )}</a>
-			${mw.msg( 'pipe-separator' )}
-			<a href="${contribsUrl}">${mw.msg( 'contribslink' )}</a>${mw.msg( 'parentheses-end' )}
-		`,
+		// Create links using the jQuery objects so they're properly escaped
```

Comment:
Just curious, what would not be properly escaped? The template literals should respect any double-quotes in interpolated expressions, and we're using `$.parseHTML` on the final markup.

Reply:
So $.parseHTML would not make sure that the actual strings would be escaped; For example, if we get some result that's `http://foo.com/?"><script>alert('boo');</script>` that would cause an xss injection.

jQuery automatically escapes and does a couple of checks when you insert string values into the values. So, it would be ideal to escape *everything* but at the very least, link properties (like href) are more crucial, which is why I changed those right there.

Reply:
The links are built with `mw.util` which does escaping anyway, no? Security is obviously important but I wonder if we're maybe going over board here. `<`, `>`, etc. aren't allowed in usernames to begin with, and the edit summary HTML comes from the API which also escapes markup.

Reply:
I don't think they're escaped with mw.util, but either way, we should *always* escape a template.

Also, why is this overboard? We have jQuery loaded anyways, and to be fair, this feels a little less cluttered than the templates especially after the later pending patches where we add quite a bunch to them, and they start getting really confusing. I think there's a place for templates, but it's okay to be a bit more vigilant when dealing with clickable things. Also, we can't be sure about what people put into usernames or comments even if it's not `>` and `<`, and escaping is always something that should to be done, regardless of anything, so we don't miss a use case. We skipped escaping the other content (I think we should, to be fair) but it was, at least, content we're a bit more sure about, but anything inside a clickable href should be a bit more higher scrutiny.

Honestly, I discovered this when I started adding the bidi encapsulation method into the templates and it started getting super confusing and cluttered.

The tempates were good when the content was clear and short, but we keep adding more and more to it, wrapping things around (bidi or security or whatever else) that some of those templates are become much less clear than a straight-up jquery object that, by merely existing, also solves a lot of potential worries about security, xss, and other issues jquery solves for us in the background.

If we already load it, and we're already using it, and it makes things a bit more organize to where we don't even need to consider whether it should or shouldn't be escaped, then why not?

Reply:
#54 helps clean up the templating some. I personally find template literals much more readable, especially for simple text content; `document.createTextNode( ' ' + mw.msg( 'parentheses-start' ) )` for instance seems quite verbose. What we have here is 20 lines of logicless code that was just 6 before. I can't imagine what the code would look like if it were a whole page of jQuery-generated content!

Anyway I've no strong feelings :) The jQuery stuff isn't what I'm questioning, rather I'm confused what we're escaping that isn't already escaped. If we can't rely on MediaWiki to do this then I think we have much bigger problems.

```
> mw.util.getUrl('Special:Contributions/<script>')
"/wiki/Special:Contributions/%3Cscript%3E"
```
Is there another example of what we're worried about? If mw.util doesn't do proper escaping, I think it certainly should.

At any rate, the code is already written, so there's no reason to change it.

Reply:
Yeah some of this came out of me trying to make sense of the templating in #54 while thinking about my new patch that needs to wrap things. Some of the new wrap functions made things more confusing rather than less, so I didn't think that changing only the links was much of a loss.

The `document.createTextNode( ' ' + mw.msg( 'parentheses-start' ) )` is only there because when I encapsulated the links, I thought it would be easier to encapsulate them all at once in a jQuery array, so Roan and I "shoved" the plaintext in there. We could, alternatively, create 3 jQuery links and combine them in a template -- which would get rid of the above. At that stage, though, it just didn't seem to matter *that* much. I could bring that back, though, if you think it looks a lot better?

In relation to the encapsulation -- I agree that in general terms, because we know exactly what we're getting from the DB and (hopefully, yeesh) from the `mw.util` methods, it's probably not insanely critical to escape things, but I feel like that is the "right way" in general when writing code for the web, so adding that security measure -- combined with the fact it's slightly more readable for the links -- felt better to me. 

I think that using templates is a matter of balance. They can make things much much much more clear, but they can also turn out to be a blob of "messy" string-in-var-in-string mess, and I think we just need to be careful of that extreme too. I started noticing this when I was looking at the other PRs and they started being a bit more complicated.

With #54 for example, I was wondering if (in the future?) we want to do something where the promise changes the DOM, which is clearer in concept (not necessary right now, but if we start adding more and more behavior, then that would make it consolidated into a single logical operation) -- and with the way that the template started to look super bloated, I was getting confused about the inline methods and the big string concatenations being full of `${xxx}` that was inside `${function(xxx)}` and near each other.

There's a general rule when building the DOM to try and avoid concatenating elements because the browser's DOM parser has its own (very good) way of preventing us from making mistakes. It's usually better to either use `document.createNode` or jQuery. That said, templates definitely have their place, but when things become dependent on a lot of sanitation/wrapping/methods I felt like it was time to re-assess.

Some of this is also personal preference over what each of us finds as "clearer" :D 
I think the best thing here is that you and I could continue to tug each other a bit, we can come up with a balance that we're both happy with -- which would likely also bring balance to unforseen issues with either approach.